File: python/code-sample/how-the-write-function-in-python-runs-as-fast-as-that-in-c/disk-io-benchmark.c

#include "stdio.h"
#include "stdlib.h"
#include "unistd.h"
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>
#include <errno.h>


int main(int argc, char * argv[]){
  int pagesize, fd, sync, last_sync, flags;
  long total_size, write_bytes;
  char * temp_buf, *file_path, path_temp[] = "/tmp/disk-io-XXXXXX";
  if(argc < 3){
    printf("total size and page size are needed.\n");
    return 0;
  }

  total_size = atol(argv[1]);
  pagesize = atoi(argv[2]);
  sync = 0;
  last_sync = 0;
  if(argc > 3) sync = atoi(argv[3]);
  if(argc > 4) last_sync = atoi(argv[4]);

  flags = O_CREAT|O_WRONLY;
  if(sync) flags |= O_SYNC;
  temp_buf = malloc(pagesize);
  memset(temp_buf, 48, pagesize);

  file_path = mktemp(path_temp);
  if(file_path == NULL){
    printf("create temp path failed.\n");
    return 0;
  }
  fd = open(file_path, flags, 0644);
  if(fd == -1){
    printf("open file failed %s.\n", strerror(errno));
    return 0;
  }

  write_bytes = 0;
  printf("write to temp file %s .\n", file_path);
  for(int i=0; i < total_size / pagesize; i++){
    write_bytes += write(fd, temp_buf, pagesize);
  }
  if(last_sync) fsync(fd);

  printf("successfully write %ld bytes.\n", write_bytes);
  return 0;
}

File: python/code-sample/how-the-write-function-in-python-runs-as-fast-as-that-in-c/disk-io-benchmark.py

"""Disk io benchmark."""

import os
import sys
import tempfile


def write_file(total_size, page_size, sync=False, last_sync=False):
    path = tempfile.mktemp(prefix='disk-io-py', dir='/tmp')
    data = b'1' * page_size
    flags = os.O_CREAT|os.O_WRONLY
    if sync:
        flags |= os.O_SYNC

    fd = os.open(path, flags, 644)
    write_bytes = 0
    for i in range(total_size // page_size):
        write_bytes += os.write(fd, data)

    if last_sync:
        os.fsync(fd)

    print('successfully write %d bytes.' % write_bytes)


def main():
    if len(sys.argv) < 3:
        print('total size and page size are needed.\n')
        return

    total_size = int(sys.argv[1])
    page_size = int(sys.argv[2])

    sync = False
    if len(sys.argv) > 3:
        sync = bool(int(sys.argv[3]))

    last_sync = False
    if len(sys.argv) > 4:
        last_sync = bool(int(sys.argv[4]))

    write_file(total_size, page_size, sync=sync, last_sync=last_sync)


if __name__ == '__main__':
    main()
